Multi GPU instructions (#267)
hyprwm/hyprland-wiki: pages/Configuring/Multi-GPU.md

# Table of contents

{{< toc >}}

# General

> If your host machine uses multiple GPUs, you may want to primarily use a GPU for
> rendering all the elements for Hyprland including windows, animations, hardware
> acceleration for certain applications, etc.

Reviewer: ```suggestion
If your host machine uses multiple GPUs, you may want to primarily use one GPU
for rendering all the elements for Hyprland including windows, animations, and
another for hardware acceleration for certain applications, etc.
```

This setup is very common in the likes of gaming laptops,

> GPU-passthrough(without VFIO) capable hosts, and if you have multiple GPUs in

Reviewer: ```suggestion
GPU-passthrough (without VFIO) capable hosts, and if you have multiple GPUs in
```

general.

> If your host machine uses multiple GPUs, you may want to primarily use a GPU for
> rendering all the elements for Hyprland including windows, animations, hardware
> acceleration for certain applications, etc.
> 
> This setup is very common in the likes of gaming laptops,
> GPU-passthrough(without VFIO) capable hosts, and if you have multiple GPUs in
> general.
> 
> 

Reviewer: ```suggestion
```
Duplicated.

# Detecting GPUs

For this case, the writer is taking the example of their laptop.

Upon running `lspci | grep -E 'VGA|3D'`, One can list all the video devices
available.

```
01:00.0 VGA compatible controller: NVIDIA Corporation TU117M [GeForce GTX
1650 Mobile / Max-Q] (rev a1) 06:00.0 VGA compatible controller: Advanced Micro
Devices, Inc. [AMD/ATI] Cezanne [Radeon Vega Series / Radeon Vega Mobile Series]

> (rev c6) ```

Reviewer: Seems like the lines got merged when formatting.

```
Here it is clear that 2 GPUs are available, the dedicated NVIDIA GTX 1650 Mobile
/ Max-Q and the integrated AMD Cezanne Radeon Vega Series GPU.

Upon running `lspci | grep -E 'VGA|3D'`, One can list all the video devices
available.

```
01:00.0 VGA compatible controller: NVIDIA Corporation TU117M [GeForce GTX
1650 Mobile / Max-Q] (rev a1) 06:00.0 VGA compatible controller: Advanced Micro
Devices, Inc. [AMD/ATI] Cezanne [Radeon Vega Series / Radeon Vega Mobile Series]
(rev c6)
```

Here it is clear that 2 GPUs are available, the dedicated NVIDIA GTX 1650 Mobile
/ Max-Q and the integrated AMD Cezanne Radeon Vega Series GPU.

> 

Reviewer: Again, duplicated.

../renderD129
```

So from the above outputs, we can match the bus IDs and determine that NVIDIA is
`card0` and AMD is `card1`.

# Telling Hyprland which GPU to use

After determining which "card" belongs to which GPU, we now have to tell
Hyprland the GPU we want to use primarily.

{{< hint type=info >}}

It is generally a good idea for laptops to use the integrated GPU as the primary
renderer as this preserves battery life and is practically indistinguishable
from using the dedicated GPU on modern systems in most cases. Hyprland can be
run on integrated GPUs just fine. The same principle applies for desktop setups
with a lower and higher power rating GPUs respectively.

{{< /hint >}}

We can do so by using `WLR_DRM_DEVICES` variable.

Add the following template to `hyprland.conf`

```ini

env = WLR_DRM_DEVICES,/dev/dri/cardN

```

For our case, we want to use `card1` primarily and use it to render stuff. 

```ini

env = WLR_DRM_DEVICES,/dev/dri/card1:/dev/dri/card0

```

Here, we tell Hyprland to set priorities. If `card1` isn't available for
whatever reason, use `card0`. So if the AMD GPU isn't available, use NVIDIA. The
colon is for setting priorities, essentially.

> You shuld now be able to use a primary GPU for lesser GPU intensive tasks 

Reviewer: ```suggestion
You shuld now be able to use an integrated GPU for light tasks 
```

# Table of contents

{{< toc >}}

# General

> If your host machine uses multiple GPUs, you may want to primarily use a GPU for rendering all the elements for Hyprland including windows, animations, hardware acceleration for certain applications, etc.

Reviewer: I think in the case of mobile systems (laptops) it's wise to say that Hyprland should run on the integrated GPU, and any heavier tasks should be offloaded to the discrete GPU.

This setup is very common in the likes of gaming laptops, GPU-passthrough(without VFIO) capable hosts, and if you have multiple GPUs in general.

# Detecting GPUs

> For this case, the writer is taking the example of their laptop. Here are some of the details for the same

Reviewer: Same... what?

> Upon running `lspci | grep VGA`, I can list all the video devices installed on this laptop

Reviewer: In the case of gaming laptops, sometimes the discrete GPU shows up as 3D controller instead of VGA controller.
```suggestion
Upon running `lspci | grep -E 'VGA|3D'`, I can list all the video devices installed on this laptop
```

```
01:00.0 VGA compatible controller: NVIDIA Corporation TU117M [GeForce GTX 1650 Mobile / Max-Q] (rev a1)
06:00.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Cezanne [Radeon Vega Series / Radeon Vega Mobile Series] (rev c6)
```

Here it is clear that the writer has 2 GPUs, the dedicated NVIDIA GTX 1650 Mobile / Max-Q and the integrated AMD Cezanne Radeon Vega Series GPU.

So this would indicate, Linux should be detecting 2 GPUs, and thus 2 "cards" in `/dev/dri/`. Let us verify this using `ls /dev/dri/`

```
by-path  card0  card1  renderD128  renderD129
```

> Here, `card0` and `card1` correspond to `renderD128` and `renderD129` respectively. So for every card you have installed on your host device the format would be something like this:

Reviewer: This is extraneous.

` cardN  renderD(128+N)`

> We now need to determine which of these `card0` and card1` correspond to the AMD and NVIDIA GPU.

Reviewer: You don't need to jump through hoops to check which device is which card. Simply `ls -l /dev/dri/by-path`, and the PCI ID will show up as a link to its respective card.

This also means you can get rid of the same-vendor section as well.

| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   52C    P8     2W /  30W |      3MiB /  4096MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      2869      G   /usr/bin/Hyprland                   1MiB |     
+-----------------------------------------------------------------------------+
```

> This is with running a browser and some other stuff. We can confidentally say NVIDIA is not doing anything and is idling.

Reviewer: I don't think this is a good way of telling whether the card runs or not, since Nvdia is blocklisted for hardware acceleration iirc. Something like `vkcube` or `glxinfo` would be more adequate.
